### ISSUE TITLE
Linq query by string with \\ symbol

### ISSUE THREAD

**Issue by KopIlya** [JasperFx/marten#737]
Hello. 
Request like this working
`var result1 = martenQueryable.Where(o => o.Name.Equals("Ivan", StringComparison.OrdinalIgnoreCase)).ToList();`

But If search parameter contains symbol `\\`, then result is 0.
`var result2 = martenQueryable.Where(o => o.Name.Equals("Domen\\Ivan", StringComparison.OrdinalIgnoreCase)).ToList();`

I was try `OrdinalIgnoreCase, CurrentCultureIgnoreCase, InvariantCultureIgnoreCase`.
I want write one query that required IgnoreCase and works with "Ivan" and "Domen\\Ivan".
What is specific about the symbol `\\`?

**Comment by mysticmind** [comment]
In your query with string `"Domen\\Ivan"`, it results in SQL like `'Name' LIKE 'Domen\Ivan'` since `\` is a escape char in C#. 

If you change your query to the following, it would work:
```
// added @ to the string so that the string containing \\ is not escaped
var result2 = martenQueryable.Where(o => o.Name.Equals(@"Domen\\Ivan", StringComparison.OrdinalIgnoreCase)).ToList();

// or use \\\\
var result2 = martenQueryable.Where(o => o.Name.Equals("Domen\\\\Ivan", StringComparison.OrdinalIgnoreCase)).ToList();
```

**Comment by KopIlya** [comment]
Nope. Both proposed options do not produce results.

**Comment by mysticmind** [comment]
If you had `\\` when you had added the doc then it would have got escaped
and stored with single `\`. You will need to take care of escape chars
during both storing and retrieval.

**Comment by KopIlya** [comment]
I understood you.
If I had `\\` then it is stored like `\\`. 
If I try save `@"Domain\Ivan"` it is also stored like `\\`.
In both cases query with `@"Domen\\Ivan"` works.

But not trivial that `"Domen\\Ivan"` while maintaining and when searching behaves differently.
This confusion is made by the postgres or by marten linq provider?

**Comment by mysticmind** [comment]
`\` being an escape character it can cause several issues if you are not handling it properly. Just try this below on a console, it will clarify things for you:
```csharp
Console.WriteLine("Domain\\Ivan"); \\ result Domain\Ivan
Console.WriteLine("Domain\\\\Ivan"); \\ result Domain\\Ivan
Console.WriteLine(@"Domain\Ivan"); \\ result Domain\Ivan
Console.WriteLine(@"Domain\\Ivan"); \\ result Domain\\Ivan
```

When records get added to db, 
if insert query had `"Domain\\Ivan"` or `@"Domain\Ivan"`, it will be stored as `Domain\\Ivan`
if insert query had `"Domain\\\\Ivan"` or `@"Domain\\Ivan"`, it will be stored as `Domain\\\\Ivan`

When searching:
you could either use `o.Name.Equals("Domain\\Ivan")` or `o.Name.Equals(@"Domain\Ivan")`

In summary, postgres or marten linq provider are working as expected and does not attribute to any issue. It is how C# works.

**Comment by KopIlya** [comment]
> When records get added to db,
if insert query had "Domen\\Ivan" or @"Domen\Ivan", it will be stored as Domen\\Ivan

Yes, it is stored as `Domen\\Ivan`

> When searching:
> you could either use o.Name.Equals("Domen\\Ivan") or o.Name.Equals(@"Domen\Ivan")

No. Search if I write `o.Name.Equals("Domen\\\\Ivan")` or `o.Name.Equals(@"Domen\\Ivan")`

And I can't understand this.

**Comment by mysticmind** [comment]
Insert query with "Domen\Ivan" will result in a compiler error i.e. Unrecognized escape sequence so it is invalid. 

If you insert `@"Domen\Ivan"`, the character following `\` is alphabet hence C# will consider the string as `"Domen\Ivan"` and the same gets added to db. If you want to run a query to fetch this record, you will need to use @"Domen\Ivan"

**Comment by KopIlya** [comment]
Consider the following code:
```javascript
BaseClass newObj = new BaseClass() {Id = "qwe", SystemId = "123", Name = @"Domen\Ivan", Version = 1};

using (var session = store.OpenSession())
{
    session.Store(newObj);
    session.SaveChanges();
}

var querySession = store.QuerySession().Query<BaseClass>();
var res1 = querySession
  .Count(o => o.Name.Equals(@"Domen\Ivan", StringComparison.OrdinalIgnoreCase)); // res1 = 0
var res2 = querySession
  .Count(o => o.Name.Equals(@"Domen\\Ivan", StringComparison.OrdinalIgnoreCase)); // res2 = 1
```
I expect `res1` to be 1 and `res2` to be 0. But the opposite is true. Isn't that an error?

**Comment by KopIlya** [comment]
Can anyone tell me if there are plans to solve this issue?

**Comment by jeremydmiller** [comment]
@KopIlya Can you add a pull request to the 1.5 branch with a failing test case?

**Comment by KopIlya** [comment]
I create pull request [https://github.com/JasperFx/marten/pull/758](url)
This is my first attempt, please tell me if I did something wrong.

**Comment by jeremydmiller** [comment]
@KopIlya @mysticmind Guys, this is a Newtonsoft issue. From my research the only "fix" is to do a Replace() on the JSON string to strip out the "\\", and that's a no go from a performance standpoint in 2.0. This does work fine if you can use "/", or if you'll do the search w/ double backslashes to mimic the Json storage. 

Sorry, but I'm closing this one.

**Comment by svver** [comment]
@jeremydmiller , I've got the same issue. Could you tell what kind of issue with Newtonsoft do you mean? I've tried to serialize and deserialize a c# object with a property containing full user domain name value  with Json.Net and everything worked fine.  
If there is a bug in Json.Net we might create an issue in their github project, i think.

**Comment by jeremydmiller** [comment]
Json.Net makes the "domain\user" be "domain\\user" in the Json. I didn't find any kind of setting to control that, and all I found on StackOverflow was to do a string.Replace() on the JSON.

**Comment by svver** [comment]
Suppose we have a class that contains a string property that holds a domain user name:
```csharp
public class LoginHolder
{
    public string Login { get; set; }
}
```
We serialize an instance of `LoginHolder` and write the resulting string to the console (Json.Net version is 10.0.1):
```csharp
string loginHolderAsJsonString = JsonConvert.SerializeObject(
   new LoginHolder { Login = @"DomainName\UserName" });
Console.WriteLine(loginHolderAsJsonString);
```
And then we get the following string: 
```
{"Login":"DomainName\\UserName"}
```
This seems correct to me, because `"DomainName\\UserName"` is a json string so the `\` symbol is eacaped with the second `\`. I still can't understand how this behaviour can cause the following query to fail.
```csharp 
querySession.Count(o => o.Login.Equals(@"DomainName\UserName", 
   StringComparison.OrdinalIgnoreCase)) 
``` 
 Maybe there are problems in how this LINQ-query is translated into PostgresQL Json field query?

**Comment by jeremydmiller** [comment]
@svver Because the Linq parsing code expects the string value within the persisted Json to be the value you're searching on. There's nothing in there that's going to automatically correct "domain\user" to "domain\\user" before issuing the actual query. 

So, to reiterate, I don't think I'm enthusiastic at all about doing a blanket `Replace(@"\\", "\")` within the command building because of the performance hit. We could do something line:

```
session.Query<LogonUser>().Where(x => x.Login.IsDomainUser("domain\username"))
```

so that it could be handled explicitly when the Linq support encounters the IsDomainUser. And actually, you could build that and plug into Marten as is w/o any changes to the Marten codebase.

**Comment by KopIlya** [comment]
I look in sorce code and undersant following. All linq-queries are translated as `NpgSqlCommand`.
The command 1
`martenQueryable.Where(o => o.Name == "Domen\\Ivan")`
looks like
`select d.data, d.id, d.mt_version from public.mt_doc_user as d where d.data ->> 'Name' = :arg0` 
The command 2
`martenQueryable.Where(o => o.Name.Equals("Domen\\Ivan"))`
looks like
`select d.data, d.id, d.mt_version from public.mt_doc_user as d where d.data ->> 'Name' LIKE :arg0`
The command 3
`martenQueryable.Where(o => o.Name.Equals("Domen\\Ivan", StringComparison.OrdinalIgnoreCase))`
looks like
`select d.data, d.id, d.mt_version from public.mt_doc_user as d where d.data ->> 'Name' ILIKE :arg0`

In Posgresql I execute all this command and whey give the same result. So, this this is not problem of Marten.
But, I am confused by the fact what `==` operator in linq translated in `NpgsqlCommand` as `=`, and `string.Equals()` operator translated as `LIKE` and with ignore case as `ILIKE`.
Why in all 3 cases doesn't use the `=` operator in `NpgsqlCommand`?

**Comment by jeremydmiller** [comment]
@KopIlya `string.Equals()` has to use `LIKE` when you're doing case insensitive searches. The `string.Equals(value)` Linq expression is a minor bug that I'd guess is caused by the ordering of strategies inside of the Linq parsing code.
